lib/rubysl/rubysl-socket/lib/rubysl/socket/foreign.rb

```diff
@@ -1,7 +1,11 @@
 module RubySL
   module Socket
     module Foreign
-      extend Rubinius::FFI::Library
+      def self.attach_function(method_name, native_name = method_name, args_types, return_type)
+        blocking = (return_type == :int || return_type == :ssize_t)
+        Truffle::POSIX.attach_function(native_name, args_types, return_type,
+                                       on: self, as: method_name, blocking: blocking)
```

Comment:
Unrelated to this PR really, but maybe we should make blocking the default, and be opt-out? Less chance of missing the annotation where required that way, and harmless if you leave it on by mistake.

Reply:
The problem is `blocking` makes assumptions and has side effects:
Side-effects:
* SIGVTALRM can be sent during the call. Most of the time this is OK, but some C code could not handle EINTR from a syscall and behave wrongly because of it.
* The Ruby Thread#status is changed to `sleep` for the duration of the call. This can be observed by other threads and is sometimes used to know whether a Thread is in a blocking call (notably in specs).

Assumptions:
* We need a way to know if the call succeeded or failed, and whether it was interrupted. Currently we assume an integer-like return value and if it's `-1` then we assume that the call failed and that `errno` contains the error code. This is true for most system calls, but for instance not for `getaddrinfo()` which has different error semantics (I need to fix that in this PR).